src/RewardsStaking.sol

```diff
@@ -0,0 +1,353 @@
+    event RewardTokenRemoved(address indexed token);
+
+    error InvalidAmount();
+    error InvalidIndex();
+    error InvalidLockType();
+    error InvalidRewardToken();
+    error AlreadyAddedRewardToken();
+    error CooldownNotFinished();
+    error PositionLocked();
+    error PositionNotExists();
+    error ZeroAddress();
+
+    constructor(IERC20 _stakingToken) Ownable(msg.sender) {
+        if (address(_stakingToken) == address(0)) revert ZeroAddress();
+        stakingToken = _stakingToken;
+        deploymentDay = block.timestamp / 1 days;
+    }
+
+    // ============ ADMIN FUNCTIONS ============
+
+    function addRewardToken(address token) external onlyOwner {
+        if (token == address(0)) revert ZeroAddress();
+        if (isRewardToken[token]) revert AlreadyAddedRewardToken();
+
+        isRewardToken[token] = true;
+        rewardTokenIndex[token] = uint8(rewardTokens.length);
+        rewardTokens.push(token);
+
+        emit RewardTokenAdded(token);
+    }
+
+    function removeRewardToken(address token) external onlyOwner {
+        if (!isRewardToken[token]) revert InvalidRewardToken();
+
+        uint256 index = rewardTokenIndex[token];
+        uint256 lastIndex = rewardTokens.length - 1;
+
+        if (index != lastIndex) {
+            address lastToken = rewardTokens[lastIndex];
+            rewardTokens[index] = lastToken;
+            rewardTokenIndex[lastToken] = uint8(index);
+        }
+
+        rewardTokens.pop();
+        delete isRewardToken[token];
+        delete rewardTokenIndex[token];
+
+        emit RewardTokenRemoved(token);
+    }
+
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+
+    // ============ STAKING FUNCTIONS ============
+
+    function stake(uint256 amount, LockType lockType, bool autoClaim) external nonReentrant whenNotPaused {
+        if (amount < MIN_STAKING_AMOUNT) revert InvalidAmount();
+
+        if (autoClaim) {
+            _claim(msg.sender);
+        }
+
+        uint256 currentDay = getCurrentDaySinceDeployment();
+        uint32 startDay = uint32(currentDay + 1); // staking starts from the next day
+        uint32 unlockDay = lockType == LockType.LIQUID ? type(uint32).max : uint32(startDay + getLockDuration(lockType));
+
+        UserData storage user = userData[msg.sender];
+        uint256 positionIndex = user.positionCount++;
+
+        userPositions[msg.sender][positionIndex] = StakingPosition({
+            amount: uint128(amount),
+            startDaySinceDeployment: startDay,
+            unlockDaySinceDeployment: unlockDay,
+            cooldownEndDaySinceDeployment: 0,
+            lockType: lockType
+        });
+
+        uint256 weight = calculateWeight(amount, lockType);
+        _updateWeightDelta(msg.sender, startDay, unlockDay, int256(weight));
+
+        totalStaked += amount;
+
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
do we only support EOA accounts now? `msg.sender` does not support meta txs or smart wallets